File: database/schemas/character.py
from pydantic import BaseModel, Field, ConfigDict
from bson import ObjectId
from typing import Optional, List, Dict


# Modelo para rolagens de dados
class DiceRoll(BaseModel):
    dice_count: int = Field(..., ge=1)  # Quantidade de dados (ex: 2 em "2d6")
    dice_sides: int = Field(..., ge=2)  # Lados do dado (ex: 6 em "2d6")
    modifier: int = 0  # Modificador (ex: +3 em "2d6+3")

    def __str__(self):
        if self.modifier == 0:
            return f"{self.dice_count}d{self.dice_sides}"
        elif self.modifier > 0:
            return f"{self.dice_count}d{self.dice_sides}+{self.modifier}"
        else:
            return f"{self.dice_count}d{self.dice_sides}{self.modifier}"


# Modelo para ataques físicos/armas
class Attack(BaseModel):
    name: str
    attack_bonus: int = 0
    damage: DiceRoll
    damage_type: str = "cortante"
    range: str = "Corpo a corpo"
    description: Optional[str] = None


# Modelo para uma magia individual
class Spell(BaseModel):
    name: str
    level: int = Field(..., ge=0, le=9)
    school: str
    description: Optional[str] = None

    # Informações de ataque/dano (para magias ofensivas)
    is_attack_spell: bool = False
    attack_bonus: Optional[int] = None
    damage: Optional[DiceRoll] = None
    damage_type: Optional[str] = None
    save_dc: Optional[int] = None
    save_ability: Optional[str] = None
    range: str = "Toque"


# Modelo para os 6 atributos básicos do D&D
class Attributes(BaseModel):
    strength: int = Field(..., ge=3, le=20)
    dexterity: int = Field(..., ge=3, le=20)
    constitution: int = Field(..., ge=3, le=20)
    intelligence: int = Field(..., ge=3, le=20)
    wisdom: int = Field(..., ge=3, le=20)
    charisma: int = Field(..., ge=3, le=20)


# Modelo para informações de raça e subraça
class RaceInfo(BaseModel):
    """Informações sobre a raça selecionada"""
    race_name: str
    race_index: str
    subrace_name: Optional[str] = None
    subrace_index: Optional[str] = None
    speed: int = 30
    size: str = "Medium"

    # Bônus de atributos combinados (raça + subraça)
    ability_bonuses: Dict[str, int] = {}  # ex: {"strength": 1, "constitution": 2}

    # Traços raciais
    racial_traits: List[str] = []
    languages: List[str] = []
    proficiencies: List[str] = []


# Modelo para informações básicas do personagem
class BasicInfo(BaseModel):
    model_config = ConfigDict(
        validate_by_name=True,  # Era allow_population_by_field_name no v1
        populate_by_name=True  # Permite usar tanto field name quanto alias
    )

    name: str = Field(..., min_length=2, max_length=50)
    race_info: RaceInfo  # ← NOVO: substitui race individual
    character_class: str = Field(..., alias="class")
    level: int = Field(default=1, ge=1, le=20)
    background: str
    alignment: Optional[str] = None


# Modelo para perícias do D&D
class Skills(BaseModel):
    # Força
    athletics: bool = False

    # Destreza
    acrobatics: bool = False
    sleight_of_hand: bool = False
    stealth: bool = False

    # Inteligência
    arcana: bool = False
    history: bool = False
    investigation: bool = False
    nature: bool = False
    religion: bool = False

    # Sabedoria
    animal_handling: bool = False
    insight: bool = False
    medicine: bool = False
    perception: bool = False
    survival: bool = False

    # Carisma
    deception: bool = False
    intimidation: bool = False
    performance: bool = False
    persuasion: bool = False


# Modelo para status/stats derivados
class Stats(BaseModel):
    hit_points: int = Field(..., ge=1)
    armor_class: int = Field(..., ge=10, le=30)
    experience_points: int = Field(default=0, ge=0)


# Modelo para sistema de combate
class Combat(BaseModel):
    attacks: List[Attack] = []


# Modelo para sistema de magias
class Magic(BaseModel):
    spellcaster: bool = False
    spellcasting_ability: Optional[str] = None
    known_spells: List[Spell] = []

    # Slots de magia por nível (1-9)
    spell_slots_1: int = 0
    spell_slots_2: int = 0
    spell_slots_3: int = 0
    spell_slots_4: int = 0
    spell_slots_5: int = 0
    spell_slots_6: int = 0
    spell_slots_7: int = 0
    spell_slots_8: int = 0
    spell_slots_9: int = 0


# Modelo completo do personagem
class Character(BaseModel):
    model_config = ConfigDict(
        arbitrary_types_allowed=True,
        validate_by_name=True,
        populate_by_name=True
    )

    id: Optional[ObjectId] = None
    user_id: ObjectId
    campaign_id: Optional[ObjectId] = None  # Referência à campanha (pode ser None se não estiver em campanha)
    basic_info: BasicInfo  # ← ATUALIZADO: agora inclui race_info com subraça
    attributes: Attributes
    skills: Skills
    stats: Stats
    combat: Combat
    magic: Magic
    calculated_stats: Optional[Dict] = None  # Estatísticas calculadas salvas no banco


# Schema para criação de personagem
class CharacterCreate(BaseModel):
    model_config = ConfigDict(
        validate_by_name=True,
        populate_by_name=True,
        arbitrary_types_allowed=True
    )

    user_id: ObjectId
    campaign_id: Optional[ObjectId] = None  # ID da campanha (opcional)
    basic_info: BasicInfo  # ← ATUALIZADO: inclui race_info
    attributes: Attributes
    skills: Skills
    stats: Stats
    combat: Combat
    magic: Magic


# Schema para resposta
class CharacterResponse(BaseModel):
    model_config = ConfigDict(
        validate_by_name=True,
        populate_by_name=True
    )

    id: str
    user_id: str
    campaign_id: Optional[str] = None
    basic_info: BasicInfo  # ← ATUALIZADO: inclui race_info
    attributes: Attributes
    skills: Skills
    stats: Stats
    combat: Combat
    magic: Magic


# Schema para listar personagens (versão resumida)
class CharacterSummary(BaseModel):
    model_config = ConfigDict(
        validate_by_name=True,
        populate_by_name=True
    )

    id: str
    user_id: str
    campaign_id: Optional[str] = None
    name: str
    race_name: str  # ← NOVO: nome da raça
    subrace_name: Optional[str] = None  # ← NOVO: nome da subraça
    character_class: str = Field(alias="class")
    level: int


# ===========================
# FUNÇÕES AUXILIARES PARA SUBRAÇAS
# ===========================

def create_race_info_from_frontend_data(race_data: dict, subrace_data: Optional[dict] = None) -> RaceInfo:
    """
    Cria RaceInfo a partir dos dados do frontend

    Args:
        race_data: Dados da raça do frontend (DndRace)
        subrace_data: Dados da subraça do frontend (DndSubrace, opcional)

    Returns:
        RaceInfo: Objeto com informações combinadas da raça e subraça
    """
    # Combinar bônus de atributos
    combined_bonuses = {}

    # Adicionar bônus da raça
    for bonus in race_data.get("ability_bonuses", []):
        ability_key = bonus["ability_score"]["index"]
        combined_bonuses[ability_key] = combined_bonuses.get(ability_key, 0) + bonus["bonus"]

    # Adicionar bônus da subraça (se existir)
    if subrace_data:
        for bonus in subrace_data.get("ability_bonuses", []):
            ability_key = bonus["ability_score"]["index"]
            combined_bonuses[ability_key] = combined_bonuses.get(ability_key, 0) + bonus["bonus"]

    # Combinar traços raciais
    racial_traits = []
    if "traits" in race_data:
        racial_traits.extend([trait["name"] for trait in race_data["traits"]])
    if subrace_data and "racial_traits" in subrace_data:
        racial_traits.extend([trait["name"] for trait in subrace_data["racial_traits"]])

    # Combinar proficiências
    proficiencies = []
    if "starting_proficiencies" in race_data:
        proficiencies.extend([prof["name"] for prof in race_data["starting_proficiencies"]])
    if subrace_data and "starting_proficiencies" in subrace_data:
        proficiencies.extend([prof["name"] for prof in subrace_data["starting_proficiencies"]])

    # Combinar linguagens
    languages = []
    if "languages" in race_data:
        languages.extend([lang["name"] for lang in race_data["languages"]])
    if subrace_data and "languages" in subrace_data:
        languages.extend([lang["name"] for lang in subrace_data["languages"]])

    return RaceInfo(
        race_name=race_data["name"],
        race_index=race_data["index"],
        subrace_name=subrace_data["name"] if subrace_data else None,
        subrace_index=subrace_data["index"] if subrace_data else None,
        speed=race_data.get("speed", 30),
        size=race_data.get("size", "Medium"),
        ability_bonuses=combined_bonuses,
        racial_traits=racial_traits,
        languages=languages,
        proficiencies=proficiencies
    )


def get_combined_ability_bonuses(race_info: RaceInfo) -> Dict[str, int]:
    """
    Retorna os bônus de atributos combinados da raça e subraça

    Args:
        race_info: Informações da raça/subraça do personagem

    Returns:
        Dict[str, int]: Dicionário com os bônus por atributo
    """
    return race_info.ability_bonuses.copy()


def format_race_display_name(race_info: RaceInfo) -> str:
    """
    Formata o nome da raça para exibição

    Args:
        race_info: Informações da raça/subraça

    Returns:
        str: Nome formatado (ex: "Elfo (Alto Elfo)" ou apenas "Humano")
    """
    if race_info.subrace_name:
        return f"{race_info.race_name} ({race_info.subrace_name})"
    return race_info.race_name

File: src/types/character.ts
export interface DiceRoll {
  dice_count: number;
  dice_sides: number;
  modifier: number;
}

export interface EquipmentItem {
  index: string;
  name: string;
  url: string;
  desc?: string[];
  equipment_category: {
    index: string;
    name: string;
    url: string;
  };
  gear_category?: {
    index: string;
    name: string;
    url: string;
  };
  cost: {
    quantity: number;
    unit: string;
  };
  weight?: number;
  
  // Campos específicos para armas
  weapon_category?: string;
  weapon_range?: string;
  category_range?: string;
  damage?: {
    damage_dice: string;
    damage_type: {
      index: string;
      name: string;
      url: string;
    };
  };
  range?: {
    normal: number;
    long: number | null;
  };
  properties?: Array<{
    index: string;
    name: string;
    url: string;
  }>;
  
  // Campos específicos para armaduras
  armor_category?: string;
  armor_class?: {
    base: number;
    dex_bonus: boolean;
    max_bonus?: number | null;
  };
  str_minimum?: number;
  stealth_disadvantage?: boolean;
  
  // Campos para itens mágicos
  rarity?: {
    name: string;
  };
  attunement?: boolean;
  is_magical?: boolean;
  
  // Campos para kits de ferramentas
  tool_category?: string;
  
  // Campos para veículos
  vehicle_category?: string;
  speed?: {
    quantity: number;
    unit: string;
  };
  
  // Campos para itens com conteúdo
  contents?: Array<{
    item: {
      index: string;
      name: string;
      url: string;
    };
    quantity: number;
  }>;
  
  // Campos especiais
  special?: string[];
  capacity?: string;
}

export interface Attack {
  name: string;
  attack_bonus?: number;
  damage: DiceRoll;
  damage_type?: string;
  range?: string;
  description?: string;
}

export interface Spell {
  name: string;
  level: number;
  school: string;
  description?: string;
  is_attack_spell?: boolean;
  attack_bonus?: number;
  damage?: DiceRoll;
  damage_type?: string;
  save_dc?: number;
  save_ability?: string;
  range?: string;
}

export interface Attributes {
  strength: number;
  dexterity: number;
  constitution: number;
  intelligence: number;
  wisdom: number;
  charisma: number;
}

export interface RaceInfo {
  race_name: string;
  race_index: string;
  subrace_name?: string;
  subrace_index?: string;
  speed: number;
  size: string;
  ability_bonuses: Record<string, number>;
  racial_traits: string[];
  languages: string[];
  proficiencies: string[];
}

export interface APIReference {
  index: string;
  name: string;
  url?: string;
}

export interface ProficiencyChoice {
  choose: number;
  type?: string;
  from?: any[]; // normalmente array de APIReference ou objetos mais complexos
}

export interface StartingEquipmentItem {
  equipment: APIReference;
  quantity: number;
}

export interface StartingEquipmentOption {
  choose: number;
  type?: string;
  from?: any[];
}

export interface MultiClassingInfo {
  requirements?: any[];
  proficiencies?: APIReference[];
}


export interface DnDClass {
  index: string;
  name: string;
  hit_die?: number;
  desc?: string[]; // descrição em array, conforme a API
  proficiency_choices?: ProficiencyChoice[];
  proficiencies?: APIReference[];
  saving_throws?: APIReference[];
  starting_equipment?: StartingEquipmentItem[];
  starting_equipment_options?: StartingEquipmentOption[];
  class_levels?: string; // endpoint (ex: "/api/classes/wizard/levels")
  subclasses?: APIReference[]; // referências para subclasses
  multi_classing?: MultiClassingInfo;
  spellcasting?: any; // estrutura complexa — deixe any ou defina explicitamente se quiser
  url?: string;

  // Allow extra fields que a API pode retornar:
  [key: string]: any;
}

// ======= Tipos de Subclass (seu formato customizado) =======

export interface FeatureItem {
  index: string;
  name: string;
  description?: string; // aqui você guarda a descrição (substituindo a url)
  url?: string; // opcional — mantive para compatibilidade
}

export interface SubclassLevel {
  level: number;
  features: FeatureItem[];
}

/**
 * Atenção: o backend recebe o campo com o nome "class" (alias).
 * Em TypeScript é válido usar `class?: APIReference`, mas se preferir
 * acesse sempre como subclass["class"] no runtime para evitar conflitos.
 */
export interface Subclass {
  index: string;
  name: string;
  class?: APIReference; // campo vindo do backend (referência à class)
  desc?: string[];
  subclass_flavor?: string;
  subclass_levels: SubclassLevel[];
  url?: string;

  [key: string]: any;
}

export interface BasicInfo {
  name: string;
  race_info: RaceInfo;
  // agora o campo "character_class" é o objeto completo da API
  character_class: DnDClass;
  level: number;
  background: string;
  alignment?: string;
}

export interface Skills {
  athletics: boolean;
  acrobatics: boolean;
  sleight_of_hand: boolean;
  stealth: boolean;
  arcana: boolean;
  history: boolean;
  investigation: boolean;
  nature: boolean;
  religion: boolean;
  animal_handling: boolean;
  insight: boolean;
  medicine: boolean;
  perception: boolean;
  survival: boolean;
  deception: boolean;
  intimidation: boolean;
  performance: boolean;
  persuasion: boolean;
}

export interface Stats {
  hit_points: number;
  armor_class: number;
  experience_points?: number;
}

export interface Combat {
  attacks: Attack[];
}

export interface Magic {
  spellcaster: boolean;
  spellcasting_ability?: string;
  known_spells: Spell[];
  spell_slots_1: number;
  spell_slots_2: number;
  spell_slots_3: number;
  spell_slots_4: number;
  spell_slots_5: number;
  spell_slots_6: number;
  spell_slots_7: number;
  spell_slots_8: number;
  spell_slots_9: number;
}

export interface CharacterDetails {
  background: string;
  alignment?: string;
  personality_traits?: string;
  ideals?: string;
  bonds?: string;
  flaws?: string;
  backstory?: string;
  appearance?: string;
}

export interface CalculatedStats {
  ability_modifiers?: Record<string, number>;
  proficiency_bonus?: number;
  spellcasting_modifier?: number;
  spell_attack_bonus?: number;
  spell_save_dc?: number;
  initiative_bonus?: number;
  race_info?: {
    display_name: string;
    race_name: string;
    subrace_name?: string;
  };
  combined_bonuses?: Record<string, number>;
  racial_traits?: string[];
  languages?: string[];
  proficiencies?: string[];
  final_attributes?: Record<string, number>;
  base_attributes?: Record<string, number>;
}

export interface Character {
  id: string;
  campaign_id?: string;
  user_id?: string;

  basic_info: BasicInfo;

  attributes: Attributes;
  skills: Skills;

  stats: Stats;
  combat: Combat;

  magic: Magic;

  details: CharacterDetails;

  equipment: EquipmentItem[];
  features: string[];
  languages: string[];
  proficiencies: string[];

  player_name?: string;
  is_active: boolean;

  created_at?: string;
  updated_at?: string;
  avatar_url?: string;

  calculated_stats?: CalculatedStats;

  // subclasse conforme o backend -> objeto completo no seu formato
  chosen_subclass?: Subclass;

  // referência leve para subclasse (opcional)
  chosen_subclass_ref?: APIReference;
}

// Schema para listagem resumida de personagens
export interface CharacterSummary {
  id: string;
  user_id: string;
  campaign_id?: string;
  name: string;
  race_name: string;
  subrace_name?: string;
  // na listagem resumida, faz sentido devolver uma referência leve (APIReference)
  class: APIReference;
  level: number;
  player_name?: string;
  is_active: boolean;
  avatar_url?: string;
}